### PR TITLE
PERF-#5573: Don't trigger axes computation in `columnarize` function

### LINKED ISSUE #5573: PERF: don't trigger axes computation in `columnarize` function
In some cases, we know in advance that the Query Compiler will contain a column and it does not need to be transposed (what `columnarize` function does, for which it materializes axes).

We can avoid materialization the way `HDK` does. With the help of an additional flag in the query compiler.

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Anatoly Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #5573 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

**Review by dchigarev — COMMENTED** [review]
left couple of notes, otherwise looks good!

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 1]
```diff
@@ -242,10 +242,13 @@ class PandasQueryCompiler(BaseQueryCompiler):
     ----------
     modin_frame : PandasDataframe
         Modin Frame to query with the compiled queries.
+    shape_hint : {"row", "column", None}, default: None
+        Shape hint for frames known to be a column or a row, otherwise None.
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py:246]: We can document this as an attribute: 
https://github.com/modin-project/modin/blob/d3948666bf211c296f3a5640f23a9471cd4f54f0/modin/core/storage_formats/base/query_compiler.py#L103

File: modin/core/storage_formats/base/query_compiler.py [hunk 2]
```diff
@@ -1068,6 +1069,9 @@ def columnarize(self):
         BaseQueryCompiler
             Transposed new QueryCompiler or self.
         """
+        if self._shape_hint == "column":
+            return self
+
```
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/core/storage_formats/base/query_compiler.py:1074]: I don't like asserts there that will trigger at least one axis.

[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/core/storage_formats/base/query_compiler.py:1074]: makes sense, let's leave it as it is.

File: modin/core/storage_formats/pandas/query_compiler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/core/storage_formats/pandas/query_compiler.py:246]: Added

**Review by dchigarev — APPROVED** [review]
Nice changes!

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 3]
```diff
@@ -2236,7 +2239,7 @@ def getitem_column_array(self, key, numeric=False):
             new_modin_frame = self._modin_frame.take_2d_labels_or_positional(
                 col_labels=key
             )
-        return self.__constructor__(new_modin_frame)
+        return self.__constructor__(new_modin_frame, shape_hint)
```
Comment (YarShev) [inline modin/core/storage_formats/pandas/query_compiler.py:2242]:
```suggestion
        return self.__constructor__(new_modin_frame, shape_hint=shape_hint)
```
If a parameter is one of the kwargs, I prefer to pass it as a keyword arg.

Comment (anmyachev) [reply modin/core/storage_formats/pandas/query_compiler.py:2242]: Do you mean if the parameter is optional?

Comment (YarShev) [reply modin/core/storage_formats/pandas/query_compiler.py:2242]: yep

**Review by YarShev — APPROVED** [review]

**Comment by anmyachev** [comment]
@dchigarev @YarShev let's merge it

**Pull request merged**